### PR TITLE
Add Liveness for PTX and Vulkan

### PR DESCRIPTION
**Pull request by tonghaining**
This PR adds liveness features and tests for PTX and Vulkan models

### CODE REVIEW

File: dartagnan/src/test/java/com/dat3m/dartagnan/litmus/LitmusPTXv7_5LivenessTest.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import com.dat3m.dartagnan.utils.rules.Providers;
+import com.dat3m.dartagnan.wmm.Wmm;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.io.IOException;
+import java.util.EnumSet;
+
+@RunWith(Parameterized.class)
+public class LitmusPTXv7_5LivenessTest extends AbstractLitmusTest {
```
Comment (hernanponcedeleon) [inline dartagnan/src/test/java/com/dat3m/dartagnan/litmus/LitmusPTXv7_5LivenessTest.java:16]: Please also add a test class for PTX v6.0

File: litmus/PTX/Manual/SL-future.litmus [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+x=0;
+m=0;
+P0:r0=0;
+P0:r1=0;
+P1:r2=0;
+P1:r3=0;
+}
+ P0@cta 0,gpu 0                   | P1@cta 1,gpu 0                   ;
+ ld.weak r0, x                    | atom.relaxed.cta.cas r2, m, 0, 1 ;
+ fence.sc.gpu                     | beq r3, r2, LC00                 ;
```
Comment (hernanponcedeleon) [inline litmus/PTX/Manual/SL-future.litmus:15]: Shouldn't this (right) be `bne`?

File: litmus/PTX/Manual/SL-cas.litmus [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+x=0;
+m=0;
+P0:r0=0;
+P1:r1=0;
+P1:r2=0;
+P1:r3=0;
+}
+ P0@cta 0,gpu 0                   | P1@cta 1,gpu 0                   ;
+ st.weak x, 1                     | atom.relaxed.cta.cas r1, m, 0, 1 ;
+ fence.sc.gpu                     | beq r1, r2, LC00                 ;
```
Comment (hernanponcedeleon) [inline litmus/PTX/Manual/SL-cas.litmus:15]: Shouldn't this (right) be `bne`?

File: dartagnan/src/test/resources/dartagnan-skip.csv [hunk 4]
```diff
@@ -157,4 +157,32 @@ litmus/RISCV/SB+popaqs+NEW.litmus
 litmus/RISCV/SB+popar+poarar+NEW.litmus
 litmus/RISCV/SB+popar+poarp+NEW.litmus
 litmus/RISCV/SB+popars+NEW.litmus
-litmus/RISCV/SB+porlps+NEW.litmus
\ No newline at end of file
+litmus/RISCV/SB+porlps+NEW.litmus
+litmus/PTX/Liveness/2_threads_4_instructions/62_simple.litmus
```
Comment (hernanponcedeleon) [inline dartagnan/src/test/resources/dartagnan-skip.csv:161]: Add comment "The following tests have spinloops which are not side-effect-free (due to the use of Exch) and thus we cannot verify their liveness"

**Comment by hernanponcedeleon** [comment]
The new test from Figures 9 and 11 are wrong. In the paper, they have this + and - indicating that some lines were added or removed. These are the correct versions for Figure 11 (please also create two tests for Figure 9 with the corresponding lines removed, also check the init value of `m`)
```
PTX SL-future+plus
"Adapted from Figure 11 in
GPU Concurrency: Weak Behaviours and Programming Assumptions
https://dl.acm.org/doi/pdf/10.1145/2694344.2694391"
{
x=0;
m=1;
P0:r0=0;
P0:r1=0;
P1:r2=0;
}
 P0@cta 0,gpu 0                 | P1@cta 1,gpu 0                   ;
 ld.weak r0, x                  | atom.relaxed.cta.cas r2, m, 0, 1 ;
 fence.sc.gpu                   | bne r2, 0, LC00                  ;
 atom.relaxed.cta.exch r1, m, 0 | fence.sc.gpu                     ;
                                | st.weak x, 1                     ;
                                | LC00:                            ;
exists
 (P0:r0 == 1 /\ P1:r2 == 0)
```
and
```
PTX SL-future-minus
"Adapted from Figure 11 in
GPU Concurrency: Weak Behaviours and Programming Assumptions
https://dl.acm.org/doi/pdf/10.1145/2694344.2694391"
{
x=0;
m=1;
P0:r0=0;
P0:r1=0;
P1:r2=0;
}
 P0@cta 0,gpu 0 | P1@cta 1,gpu 0                   ;
 ld.weak r0, x  | atom.relaxed.cta.cas r2, m, 0, 1 ;
 st.weak m, 0   | bne r2, 0, LC00                  ;
 fence.sc.gpu   | st.weak x, 1                     ;
                | LC00:                            ;
exists
 (P0:r0 == 1 /\ P1:r2 == 0)
```

File: dartagnan/src/test/java/com/dat3m/dartagnan/litmus/LitmusPTXv7_5LivenessTest.java (hunk 1 above)
Comment (tonghaining) [reply dartagnan/src/test/java/com/dat3m/dartagnan/litmus/LitmusPTXv7_5LivenessTest.java:16]: Added

**Comment by hernanponcedeleon** [comment]
@tonghaining can you add **all** CADP test (i.e., not only the ones that we run for liveness) to the Vulkan-CK-expected so be sure that all tests have the correct syntax?

**Comment by tonghaining** [comment]
> @tonghaining can you add **all** CADP test (i.e., not only the ones that we run for liveness) to the Vulkan-CK-expected so be sure that all tests have the correct syntax?

It can't check all the CADP tests because those with `exchange` will return 'UNKNOWN'

**Pull request merged**